network/ipfs_impl.go

```diff
@@ -172,18 +176,26 @@ func (bsnet *impl) Provide(ctx context.Context, k cid.Cid) error {
 func (bsnet *impl) handleNewStream(s inet.Stream) {
 	defer s.Close()
 
+	bsnet.receiverLk.RLock()
```

Comment:
Let's just stash a copy of the receiver here. That way we don't have to keep re-locking (it shouldn't change, right?).

Reply:
that makes sense